allennlp/training/model.py

```diff
@@ -0,0 +1,94 @@
+        """
+        Defines the forward pass of the model. In addition, to facilitate easy training,
+        this method is designed to compute a loss function defined by a user.
+
+        The input dictionary is comprised of everything required to perform a
+        training update, `including` labels. It is down to the user to ensure
+        that inference can be performed (using compute_loss = False) without the
+        presence of these labels! Hence, any inputs not available at inference
+        time should only be used inside an ``if compute_loss:`` conditional block.
+
+        The intended sketch of this method is as follows:
+
+        >>> def forward(inputs, compute_loss=False):
+        >>>     ....
+        >>>     ....
+        >>>     outputs = self.last_layer(inputs)
+        >>>
+        >>>     if compute_loss:
+        >>>         loss = self.compute_loss(outputs, targets, model_state)
+        >>>         return outputs, loss
+        >>>     else:
+        >>>         return outputs, None
+
+        Parameters
+        ----------
+        inputs: Dict[str, torch.Tensor], required.
+            A dictionary of tensors comprising everything needed to perform a
+            training update, `including` labels. At inference time, simply
+            pass an input dictionary which does not include labels and
+        compute_loss: bool, optional (default = False)
+            If true, this method should call :func:`compute_loss` and return
+            both the model outputs and a scalar loss to optimise.
+
+        Returns
+        -------
+
+        outputs: torch.Tensor or List[torch.Tensor]
+            The outputs from the model.
+        loss: float, optional
+            If ``compute_outputs=True``, return a scalar loss to optimise.
+
+        """
+
+        raise NotImplementedError
+
+    def compute_loss(self,
+                     model_output: Union[torch.Tensor, List[torch.Tensor]],
+                     targets: Union[torch.Tensor, List[torch.Tensor]],
+                     model_state: Dict[str, Any] = None) -> float:
+        """
+        Computes a scalar loss function to optimise. This method is designed
+        to be called and returned by :func:`forward` when `compute_loss = True`.
+        It is part of the public interface however, as there are scenarios in which
+        direct evaluation may be useful, such as during model comparison.
+
+        Parameters
+        ----------
+        model_output : torch.Tensor or List[torch.Tensor], required.
+            The output of the result of calling :func:`forward`. This
+            can be a list, as complex models may have many outputs.
+        targets : torch.Tensor or List[torch.Tensor], required.
+            The gold targets for computing some loss function with respect to
+            the model outputs.
+        model_state : Dict[str, Any], optional (default = None)
```

Comment:
This could be the actual inputs to the model in a case where you are doing semi-supervised learning where for some of your data you have additional inputs, like POS tags etc, for which you want to augment/change your loss.